File: csharp/day03.cs
using System;
using System.IO;
using System.Text.RegularExpressions;

namespace AdventOfCode2020
{
    public class Day03
    {
        public static void Execute()
        {
            var lines = File.ReadAllLines("../../../../input/input03.txt");
            var width = lines[0].Length;
            var height = lines.Length;
            ulong Slide(int dy, int dx)
            {
                int x = 0, y = 0;
                var treesEncountered = 0ul;
                while (y < height - dy)
                {
                    y += dy;
                    x = (x + dx) % width;
                    if (lines[y][x] == '#')
                    {
                        ++treesEncountered;
                    }
                }
                return treesEncountered;
            }

            Console.WriteLine($"Part 1: {Slide(1, 3)}");
            Console.WriteLine($"Part 2: {Slide(1, 1) * Slide(1, 3) * Slide(1, 5) * Slide(1, 7) * Slide (2, 1)}");
            Console.WriteLine();
        }
    }
}

File: 2020/python/day03.py
from utils import get_input_path

day = 3
print(f"Day {day:02}")
with open(get_input_path(day)) as f:
    lines = f.readlines()
    width = len(lines[0]) - 1
    height = len(lines)


def slide(dy: int, dx: int) -> int:
    x, y, trees_encountered = 0, 0, 0
    while y < height - dy:
        y += dy
        x = (x + dx) % width
        if lines[y][x] == "#":
            trees_encountered += 1
    return trees_encountered


print("Part 1:", slide(1, 3))
print("Part 2:", slide(1, 1) * slide(1, 3) * slide(1, 5) * slide(1, 7) * slide(2, 1))
